samtranslator/model/sam_resources.py

```diff
@@ -1848,28 +1848,62 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
         verify_profile_variables_replaced(profile_properties)
 
         generated_resources: List[Resource] = []
+
         if profile_type == "AWS_IAM_ROLE_MANAGED_POLICY":
             generated_resources.append(
-                self._construct_iam_policy(source, destination, profile_properties, resource_resolver)
+                self._construct_iam_policy(
+                    source, destination, profile_properties, resource_resolver, dest_index, multi_dest
+                )
+            )
+        elif profile_type == "AWS_SQS_QUEUE_POLICY":
+            generated_resources.append(
+                self._construct_sqs_queue_policy(source, destination, profile_properties, dest_index, multi_dest)
+            )
+        elif profile_type == "AWS_SNS_TOPIC_POLICY":
+            generated_resources.append(
+                self._construct_sns_topic_policy(source, destination, profile_properties, dest_index, multi_dest)
             )
-        if profile_type == "AWS_SQS_QUEUE_POLICY":
-            generated_resources.append(self._construct_sqs_queue_policy(source, destination, profile_properties))
-        if profile_type == "AWS_SNS_TOPIC_POLICY":
-            generated_resources.append(self._construct_sns_topic_policy(source, destination, profile_properties))
-        if profile_type == "AWS_LAMBDA_PERMISSION":
+        elif profile_type == "AWS_LAMBDA_PERMISSION":
             generated_resources.extend(
-                self._construct_lambda_permission_policy(source, destination, profile_properties)
+                self._construct_lambda_permission_policy(
+                    source, destination, profile_properties, dest_index, multi_dest
+                )
             )
+        else:
+            raise InvalidResourceException(self.logical_id, f"Profile type {profile_type} is not supported")
+        return generated_resources
+
+    @cw_timer
+    def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
+        resource_resolver: ResourceResolver = kwargs["resource_resolver"]
+        original_template = kwargs["original_template"]
 
-        generated_logical_ids = [resource.logical_id for resource in generated_resources]
+        multi_dest = True
+        if isinstance(self.Destination, dict):
+            multi_dest = False
+            self.Destination = [self.Destination]
+
+        list_generated_resources: List[Resource] = []
+
+        for dest_index, dest in enumerate(self.Destination):
+            try:
+                destination = get_resource_reference(dest, resource_resolver, self.Source)
+                source = get_resource_reference(self.Source, resource_resolver, dest)
+            except ConnectorResourceError as e:
+                raise InvalidResourceException(self.logical_id, str(e))
+
+            generated_resource = self.generate_resource(source, destination, dest_index, multi_dest, resource_resolver)
```

Comment:
This should be _resources_, right?

Reply:
yes.

Reply:
`generated_resource` seems right here. It's passing only this specific source and destination pair.

Reply:
For lambda permission with `READ + WRITE`, we creates 2 lambda permissions. In other cases, we only create 1 policy